django/db/models/base.py

```diff
@@ -1333,12 +1333,19 @@ def _get_next_or_previous_in_order(self, is_next):
             setattr(self, cachename, obj)
         return getattr(self, cachename)
 
-    def _get_field_value_map(self, meta, exclude=None):
+    def _wrap_value(self, value, field):
+        if hasattr(value, "resolve_expression"):
+            value.output_field = field
```

Comment:
Instead of doing this here we could set `output_field` when the `DatabaseDefault` is first instantiated

Reply:
Were you thinking of something like

```diff
diff --git a/django/db/models/expressions.py b/django/db/models/expressions.py
index ffb9f3c816..5c33f6c6b2 100644
--- a/django/db/models/expressions.py
+++ b/django/db/models/expressions.py
@@ -1252,7 +1252,37 @@ def as_sql(self, compiler, connection):
 class DatabaseDefault(Expression):
     """Placeholder expression for the database default in an insert query."""

+    def __init__(self, expression, output_field=None):
+        super().__init__(output_field)
+        self.expression = expression
+
+    def get_source_expressions(self):
+        return [self.expression]
+
+    def set_source_expressions(self, exprs):
+        (self.expression,) = exprs
+
+    def resolve_expression(
+        self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False
+    ):
+        resolved_expression = self.expression.resolve_expression(
+            query=query,
+            allow_joins=allow_joins,
+            reuse=reuse,
+            summarize=summarize,
+            for_save=for_save,
+        )
+        # Defaults used outside of INSERTs context should resolved to
+        # their underlying expression.
+        if not for_save:
+            return resolved_expression
+        return DatabaseDefault(
+            resolved_expression, output_field=self._output_field_or_none
+        )
+
     def as_sql(self, compiler, connection):
+        if not connection.features.supports_default_keyword_in_insert:
+            return compiler.compile(self.expression)
         return "DEFAULT", []


diff --git a/django/db/models/fields/__init__.py b/django/db/models/fields/__init__.py
index 796c4d23c4..d1f31f0211 100644
--- a/django/db/models/fields/__init__.py
+++ b/django/db/models/fields/__init__.py
@@ -983,13 +983,7 @@ def get_internal_type(self):

     def pre_save(self, model_instance, add):
         """Return field's value just before saving."""
-        value = getattr(model_instance, self.attname)
-        if not connection.features.supports_default_keyword_in_insert:
-            from django.db.models.expressions import DatabaseDefault
-
-            if isinstance(value, DatabaseDefault):
-                return self._db_default_expression
-        return value
+        return getattr(model_instance, self.attname)

     def get_prep_value(self, value):
         """Perform preliminary non-db specific value checks and conversions."""
@@ -1031,7 +1025,9 @@ def _get_default(self):
         if self.db_default is not NOT_PROVIDED:
             from django.db.models.expressions import DatabaseDefault

-            return DatabaseDefault
+            return lambda: DatabaseDefault(
+                self._db_default_expression, output_field=self
+            )

         if (
             not self.empty_strings_allowed
```

Reply:
I wasn't thinking of that but it's worth considering.  Because this overloads the responsibility of the class we just need to make sure we're ok the adaptive behaviour? 🤔 My initial thoughts are that it seems to fit nicely.

Reply:
Also, I've been using the raw `db_default` when I should've been using `_db_default_expression` - I completely forgot about this 😅

Reply:
Ok I tried this approach and it's better because we're consolidating into single-point-of-control 👍